cmd/routed-eni-cni-plugin/cni.go

```diff
@@ -434,14 +417,58 @@ func del(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
 	return nil
 }
 
-func cleanUpPodENI(podVlanId int, log logger.Logger, containerId string, driverClient driver.NetworkAPIs) error {
-	err := driverClient.TeardownPodENINetwork(podVlanId, log)
-	if err != nil {
-		log.Errorf("Failed on TeardownPodNetwork for container ID %s: %v",
-			containerId, err)
-		return errors.Wrap(err, "del cmd: failed on tear down pod network")
+// tryDelWithPrevResult will try to process CNI delete request without IPAMD.
+// returns true if the del request is handled.
+func tryDelWithPrevResult(driverClient driver.NetworkAPIs, conf *NetConf, k8sArgs K8sArgs, contVethName string, log logger.Logger) (bool, error) {
+	// prevResult might not be available, if we are still using older cni spec < 0.4.0.
+	prevResult, ok := conf.PrevResult.(*current.Result)
+	if !ok {
+		return false, nil
 	}
-	return nil
+
+	dummyIfaceName := generateHostVethName(dummyVlanInterfacePrefix, string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
+	_, dummyIface, found := findCNIInterfaceByName(prevResult.Interfaces, dummyIfaceName)
+	if !found {
+		return false, nil
+	}
+	podVlanID, err := strconv.Atoi(dummyIface.Mac)
+	if err != nil || podVlanID == 0 {
+		return true, errors.Errorf("malformed vlanID from prevResult: %s", dummyIface.Mac)
+	}
+
+	containerIfaceIndex, _, found := findCNIInterfaceByName(prevResult.Interfaces, contVethName)
+	if !found {
+		return false, nil
```

Comment:
Should we return "nil" here? It should be a hard failure right? Since we have the previous result but not able to find the interface.

Reply:
i want to fallback to old code if anything not expected here (thus it retains original behavior).

But i think it make sense to hard fail here as well, as cni spec should guarantee that.

Reply:
change it to hard fail :D

Reply:
How will this work for upgrade cases if we hard fail here? Let's say, we created an SGPP pod with say v1.10.1, then we will not have the interface information filled up during add operation. Now let's say we move to v1.10.2(with cni spec 0.4.0) - and we now delete the SGPP pod, we will now receive `PrevResult` but the interface will be missing isn't it? and we should fall back to previous sgpp delete code flow. Am I missing something here?